### DISCUSSION TITLE
Shared (static) variables

### DISCUSSION THREAD

**Discussion by PTKu** [simatic-ax/.discussions#15] [General]
Hi AX team! 

Is there a way to declare something known as shared members? There is a `VAR END_VAR` section that can declare variables that in the terminology of AX are variables that are shared among multiple cycles. 

We would, however, need to have variables that would be shared among the type of the same class. These are known as `static` in C# or shared in VB.

Thanks!

**Comment by BeckerStS** [comment]
Hi Peter,

not 100% sure about your concrete usecase, but the following will probably not solve it I assume:
`VAR (CONSTANT) PUBLIC`
`...`
`END_VAR`

@theBadT  
I guess he is asking for something like that:
`VAR_EXTERNAL (CONSTANT)`
`...`
`END_VAR`
... but for classes ... which unfortunatly will lead to implicit, undeclared dependencies to global declarations.
For class-methods although this declaration is supported (and may solve your usecse @PTKu ?).
Thomas, do u have an other idea ?

Greetings

**Comment by PTKu** [comment]
Thanks @BeckerStS 
Constant would do for primitive types, but we would need shared access to a complex type (CLASS). The problem with EXTERNAL is that it cannot be declared in a CLASS. Maybe it is possible to declare a shared variable in a PROGRAM, but again it cannot be declared in a library-type project.

The use case is the following. We would like to prevent NULL access in certain cases by returning a dummy object instead of NULL. This approach would prevent the PLC program from going wild when an object is not properly initialized (we would handle the error differently without stopping the PLC).

Here we have a definition of a dummy object
https://github.com/ix-ax/ix.framework/blob/f25efc14767703a0ea5d28f2eaca3fa0d8d589fc/src/core/ctrl/src/AxoObject/AxoObject.st#L21

And this is how an uninitialized object will return it:

https://github.com/ix-ax/ix.framework/blob/f25efc14767703a0ea5d28f2eaca3fa0d8d589fc/src/core/ctrl/src/AxoObject/AxoObject.st#L81

The problem with this approach is that we are polluting all objects with the dummy instance. This is why we would like to have some fixed reference (a shared dummy object) to return when necessary.

CoDeSys has VAR_STAT that behaves like `static` in C. 
https://help.codesys.com/api-content/2/codesys/3.5.14.0/en/_cds_vartypes_var_stat/

There may be some other way to address this issue, but we would like to avoid adding parameters to initialization, additional method calls, or assignments.

**Comment by sjuergen** [comment]
The IEC61131-3 does not support that feature. But CodeSys states that they have extended the IEC. So the question is: are there some improvements planned for the IEC in a newer version? @theBadT  @drsoran 


Static members/methods, properties, constructors (instead of initializers) and delegates could improve the ST language itself.

**Reply by PTKu** [reply]
It indeed would. Something to consider is if strict adherence to IEC61131-3 is a limiting factor. IMHO, it is. IEC61131-3 releases every ten (or more) years and is in the oblivion of actors halting progress. It is good to follow the standard, but extending here and there where it makes sense would not be a bad thing. Then IEC can take inspiration for the next edition😁. To my knowledge, there are no leaps in the upcoming release of the standard.
